### PR TITLE
Upgrade to Android NDK r20 [ci full]

### PR DESCRIPTION
**Pull request by eoger**
Fixes https://github.com/mozilla/application-services/issues/1029.
Fixes https://github.com/mozilla/application-services/issues/1913.
Fixes https://github.com/mozilla/application-services/issues/1912.

This will need thorough testing as it might break stuff subtly on different Android architectures.

### CODE REVIEW

**Comment by eoger** [comment]
Asking for review: Even though this is not required right now to unblock people's work, I believe we could avoid a lot of pain in the future and pay a little bit of technical debt by using a newer NDK.

File: docs/howtos/setup-android-build-environment.md [hunk 1]
```diff
@@ -21,18 +21,15 @@ At the end of this process you should have the following environment variables s
...
 - `JAVA_HOME`
 
 These variables are required every time you build, so you should add them to
 a rc file or similar so they persist between reboots etc.
 
-1. Install NDK r15c from https://developer.android.com/ndk/downloads/older_releases
-   (yes, this sucks, but newer versions don't understand the `--deprecated-headers`
-   argument required to build OpenSSL against a v21 toolchain).
-    - Extract it, put it somewhere (`$HOME/.android-ndk-r15c` is a reasonable
+1. Install NDK r20 from https://developer.android.com/ndk/downloads/older_releases
```
Comment (thomcc) [inline docs/howtos/setup-android-build-environment.md]:
```suggestion
1. Install NDK r20 from https://developer.android.com/ndk/downloads
```

File: docs/howtos/setup-android-build-environment.md [hunk 2]
```diff
@@ -21,18 +21,15 @@ At the end of this process you should have the following environment variables s
...
+    - Extract it, put it somewhere (`$HOME/.android-ndk-r20` is a reasonable
       choice, but it doesn't matter), and set `ANDROID_NDK_ROOT` to this location.
     - Set `ANDROID_NDK_HOME` to match `ANDROID_NDK_ROOT`, for compatibility with
```
Comment (thomcc) [inline docs/howtos/setup-android-build-environment.md]: Do we still need NDK_HOME as well?

File: libs/build-all.sh [hunk 3]
```diff
@@ -104,6 +104,21 @@ diff -r 65efa74ef84a coreconf/config.gypi
...
+# Early return hack to prevent NSPR Android setup
+# which does not work with ndk unified headers and clang.
+echo $'\
+@@ -2662,6 +2662,9 @@
+
+ case "$target" in
+ *-android*|*-linuxandroid*)
++    $as_echo "#define ANDROID 1" >>confdefs.h
++    ;;
++    bobotron)
```
Comment (thomcc) [inline libs/build-all.sh]: bobotron returns

(Did you mean to delete this? Is it somehow being used?)

File: libs/verify-android-environment.sh [hunk 4]
```diff
@@ -20,7 +20,7 @@ fi
 
 if [[ -z "${ANDROID_NDK_ROOT}" ]]; then
   echo "Could not find Android NDK:"
-  echo 'Please install the Android NDK r15c and then set ANDROID_NDK_ROOT.'
+  echo 'Please install Android NDK r20 and then set ANDROID_NDK_ROOT.'
```
Comment (thomcc) [inline libs/verify-android-environment.sh]: Should we check that it's r20 somehow? Seems very likely people will miss the memo to update this...

Probably worth sending out an email too, tbh.

File: docs/howtos/setup-android-build-environment.md (hunk 2 above)
Comment (eoger) [reply docs/howtos/setup-android-build-environment.md]: Yup `ANDROID_NDK_HOME` is still needed by gradle I believe.

File: libs/build-all.sh (hunk 3 above)
Comment (eoger) [reply libs/build-all.sh]: It's actually used so we don't execute the rest of the `case android` code, there's no way to early return from a `case` statement in autoconf scripts :/

Comment (eoger) [reply libs/build-all.sh]: I could rename `bobotron` to `unreacheable` to make it clearer

File: libs/verify-android-environment.sh (hunk 4 above)
Comment (eoger) [reply libs/verify-android-environment.sh]: We actually check a bit further down the file :)

Comment (eoger) [reply libs/verify-android-environment.sh]: Good idea on the email!

**Pull request merged**